### DISCUSSION TITLE
ImageMagick Binary versus Portable

### DISCUSSION THREAD

**Discussion by DadeMurphy72** [ImageMagick/ImageMagick#7454] [Help] [answered]
### ImageMagick version

7.1.1-34-portable-Q16-x64

### Operating system, version and so on

Windows 10, 11

### Description

When building an app using ImageMagick Portable, I'm assuming I do not need to install ImageMagick Binary on my developer machine, is this correct?

**Comment by dlemstra** [comment]
That is correct. You don't need to install ImageMagick when you are using the Portable installation.

**Comment by DadeMurphy72** [comment]
I guess I'm still confused. I continue to get "Failed to import Image from wand: MagickWand shared library not found". So I wrote a basic test script and spec file, all with the same results. I've tried many different ways and at the moment, it's most liklely overkill. I've completely uninstalled MagickWand and any paths or environment variables. Any idea what I'm doing wrong?

Here is they py file:
```
import os
import sys

# Set MAGICK_HOME environment variable to use local ImageMagick files
def resource_path(relative_path):
    """ Get absolute path to resource, works for dev and for PyInstaller """
    try:
        base_path = sys._MEIPASS
    except AttributeError:
        base_path = os.path.abspath(".")
    return os.path.join(base_path, relative_path)

if getattr(sys, 'frozen', False) and hasattr(sys, '_MEIPASS'):
    os.chdir(sys._MEIPASS)

# Set local paths for ImageMagick using resource_path
magick_home = resource_path("resources/ImageMagick")
ghostscript_path = resource_path("resources/ghostscript/bin")

# Set up ImageMagick environment variables
os.environ['MAGICK_HOME'] = magick_home
os.environ['MAGICK_CONFIGURE_PATH'] = magick_home
os.environ['PATH'] = magick_home + os.pathsep + ghostscript_path + os.pathsep + os.environ['PATH']
os.environ['WAND_MAGICK_LIBRARY_PATH'] = magick_home  # Point to the directory

# Debug: Print environment variables and check if binaries are accessible
print(f"MAGICK_HOME: {os.environ.get('MAGICK_HOME')}")
print(f"PATH: {os.environ.get('PATH')}")
print(f"WAND_MAGICK_LIBRARY_PATH: {os.environ.get('WAND_MAGICK_LIBRARY_PATH')}")
print(f"Contents of MAGICK_HOME directory: {os.listdir(magick_home)}")

# Log for debugging purposes
debug_log_path = os.path.join(os.path.abspath("."), "debug_log.txt")
with open(debug_log_path, "w") as log_file:
    log_file.write(f"MAGICK_HOME: {os.environ.get('MAGICK_HOME')}\n")
    log_file.write(f"PATH: {os.environ.get('PATH')}\n")
    log_file.write(f"WAND_MAGICK_LIBRARY_PATH: {os.environ.get('WAND_MAGICK_LIBRARY_PATH')}\n")
    log_file.write(f"Contents of MAGICK_HOME directory: {os.listdir(magick_home)}\n")

# Import wand after setting the environment variables
try:
    from wand.image import Image as WandImage
    print("Wand imported successfully.")
except ImportError as e:
    with open(debug_log_path, "a") as log_file:
        log_file.write(f"Failed to import Image from wand: {e}\n")
    print(f"Failed to import Image from wand: {e}")
    sys.exit(1)

```
Here is my spec file:
```
# -*- mode: python ; coding: utf-8 -*-

block_cipher = None

a = Analysis(
    ['test_wand.py'],
    pathex=['D:/Projects/PJC13'],
    binaries=[
        ('resources/ImageMagick/*.exe', 'resources/ImageMagick'),
        ('resources/ghostscript/bin/*.*', 'resources/ghostscript/bin'),
    ],
    datas=[
        ('resources/ImageMagick/*', 'resources/ImageMagick'),
        ('resources/ghostscript/bin/*', 'resources/ghostscript/bin'),
    ],
    hiddenimports=[
        'wand.image',
    ],
    hookspath=[],
    runtime_hooks=[],
    excludes=[],
    win_no_prefer_redirects=False,
    win_private_assemblies=False,
    cipher=block_cipher,
    noarchive=False,
)

pyz = PYZ(a.pure, a.zipped_data, cipher=block_cipher)

exe = EXE(
    pyz,
    a.scripts,
    [],
    exclude_binaries=True,
    name='test_wand',
    debug=False,
    bootloader_ignore_signals=False,
    strip=False,
    upx=True,
    upx_exclude=[],
    runtime_tmpdir=None,
    console=True,  # Enable console to see print statements
    icon=None,
)

coll = COLLECT(
    exe,
    a.binaries,
    a.zipfiles,
    a.datas,
    strip=False,
    upx=True,
    upx_exclude=[],
    name='test_wand',
)

```

**Comment by dlemstra** [comment]
Your software probably needs a dynamic build of ImageMagick and the portable build is static.

**Reply by DadeMurphy72** [reply]
Still at a loss. Here is where I'm at:

ImageMagick is unable to convert an Adobe Illustrator (.ai) file to a PNG format, despite Ghostscript being able to perform similar conversions independently.

Environment Setup:
ImageMagick Version: 7.1.1-34-Q16-HDRI-x65.dll (I installed, extracted the files and copied them to my project folder. I then uninstalled Imagemagick. Unfortunately, it appears I cannot use the portable version.
Ghostscript Version: 10.03.1 x64
Environment Variables Set to project folder:
MAGICK_HOME: D:\Projects\PJC13\resources\ImageMagick7
MAGICK_CODER_MODULE_PATH: D:\Projects\PJC13\resources\ImageMagick7\modules\coders
GS_PROG: D:\Projects\PJC13\resources\ghostscript\bin\gswin64c.exe
PATH: Includes the above paths

While testing a direct Ghostscript conversion, I was able to convert the AI file to PNG successfully using the command:

`D:\Projects\PJC13\resources\ghostscript\bin\gswin64c.exe -q -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -sstdout=%stderr -sDEVICE=pngalpha -dTextAlphaBits=4 -dGraphicsAlphaBits=4 -r300x300 -dUseCIEColor -sOutputFile=C:\temp\debug_output.png -f C:\temp\debug.ai -c showpage -c quit`

When using magick command for conversion:
`magick -verbose -debug all "C:\temp\debug.ai" "C:\temp\debug.png"`
I received the error "no images defined 'C:\temp\debug.png' @ error/operation.c/CLINoImageOperator/4983."

I verified that the delegates.xml and policy.xml files are correctly configured to use Ghostscript.
Verified that the correct paths are set for ImageMagick and Ghostscript.
Checked the output of magick -list configure to ensure that delegates are correctly listed, including gslib for Ghostscript.

Tried the deprecated convert command, which also failed with a similar error.

Ghostscript works correctly on its own.
ImageMagick fails to recognize and process the AI file, resulting in an error stating no images are defined.
I was able to use ImageMagick to convert a jpg to png.

Delegates configuration in ImageMagick lists Ghostscript, indicates that ImageMagick should theoretically be able to use Ghostscript for conversion.

ImageMagick is not utilizing Ghostscript correctly to read and convert the AI file, despite correct setup and environment variables. I've tried everything I can think of but obviously, I'm missing something.

**Reply by fmw42** [reply]
Post exactly what you get from 

`magick -version`

showing your Delegates.

Post your AI file

**Comment by snibgo** [comment]
> I installed, extracted the files and copied them to my project folder. I then uninstalled Imagemagick.

Huh? Why did you do that? Why not simply install IM to any desired directory? I suspect that is the problem. *Portable* versions of IM don't need to be installed, and can be copied to anywhere, and will still work. Other versions need to be installed, which writes stuff to the Windows repository. Moving an installed version to another directory will cause problems.

**Reply by DadeMurphy72** [reply]
When I install ImageMagick, the environment variables point to it and the RIP I am using this software with also uses ImageMagick. It breaks the RIP.

I did try the portable version and it will not convert/read ai files properly, even though I have all delegates and environment variables within my script, pointing to it. I can try it again, but one of the responses I received earlier was: "Your software probably needs a dynamic build of ImageMagick and the portable build is static."

I know I can create a bat file and run my application within a wrapped environment, but it is prohibiting due to being a batch file instead of an exe. (ico files as well as pinning to taskbar).

**Reply by DadeMurphy72** [reply]
My current setup is:
(new_ENV) D:\Projects\PJC13\resources\ImageMagick7>magick --version
Version: ImageMagick 7.1.1-34 Q16-HDRI x64 b0b7b17:20240623 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Channel-masks(64-bit) Cipher DPC HDRI Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png ps raqm raw rsvg tiff webp xml zip zlib
Compiler: Visual Studio 2022 (194033811)

My delegates is: 

```
<?xml version="1.0"?>
<delegatemap>
    <delegate decode="eps" command="&quot;D:\Projects\PJC13\resources\ghostscript\bin\gswin64c.exe&quot; -q -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -sstdout=%stderr -sDEVICE=pngalpha -dTextAlphaBits=4 -dGraphicsAlphaBits=4 -r%[resolution.x]x%[resolution.y] -dUseCIEColor -sOutputFile=%o -f &quot;%i&quot; -c showpage -c quit"/>
    <delegate decode="ai" command="&quot;D:\Projects\PJC13\resources\ghostscript\bin\gswin64c.exe&quot; -q -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -sstdout=%stderr -sDEVICE=pngalpha -dTextAlphaBits=4 -dGraphicsAlphaBits=4 -r%[resolution.x]x%[resolution.y] -dUseCIEColor -sOutputFile=%o -f &quot;%i&quot; -c showpage -c quit"/>
    <delegate decode="pdf" command="&quot;D:\Projects\PJC13\resources\ghostscript\bin\gswin64c.exe&quot; -q -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -sstdout=%stderr -sDEVICE=pngalpha -dTextAlphaBits=4 -dGraphicsAlphaBits=4 -r%[resolution.x]x%[resolution.y] -dUseCIEColor -sOutputFile=%o -f &quot;%i&quot; -c showpage -c quit"/>
    <delegate decode="ps:alpha" command="&quot;D:\Projects\PJC13\resources\ghostscript\bin\gswin64c.exe&quot; -q -dQUIET -dSAFER -dBATCH -dNOPAUSE -sstdout=%stderr -sDEVICE=pngalpha -dTextAlphaBits=4 -dGraphicsAlphaBits=4 -r%[resolution.x]x%[resolution.y] -dUseCIEColor -sOutputFile=%o -f - -c showpage -c quit"/>
    <delegate decode="ps:alpha" stealth="True" command="&quot;%GS_PROG%&quot; -q -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -sstdout=%stderr -sDEVICE=pngalpha -dTextAlphaBits=4 -dGraphicsAlphaBits=4 -r%[resolution.x]x%[resolution.y] -dUseCIEColor -sOutputFile=%o -f %i -c showpage -c quit"/>
</delegatemap>

```
My policy is:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE policymap [
<!ELEMENT policymap (policy)*>
<!ATTLIST policymap xmlns CDATA #FIXED "">
<!ELEMENT policy EMPTY>
<!ATTLIST policy xmlns CDATA #FIXED "">
<!ATTLIST policy domain NMTOKEN #REQUIRED>
<!ATTLIST policy name NMTOKEN #IMPLIED>
<!ATTLIST policy pattern CDATA #IMPLIED>
<!ATTLIST policy rights NMTOKEN #IMPLIED>
<!ATTLIST policy stealth NMTOKEN #IMPLIED>
<!ATTLIST policy value CDATA #IMPLIED>
]>
<policymap>
  <policy domain="resource" name="width" value="10KP"/>
  <policy domain="resource" name="height" value="10KP"/>
  <policy domain="coder" rights="read|write" pattern="AI" />
  <policy domain="coder" rights="read|write" pattern="PDF" />
  <policy domain="coder" rights="read|write" pattern="PS" />
  <policy domain="coder" rights="read|write" pattern="EPS" />
  <policy domain="path" rights="read|write" pattern="*" />
  <policy domain="module" rights="read|write" pattern="ghostscript" />
</policymap>

```

**Reply by DadeMurphy72** [reply]
I tried uploading the ai file and it won't upload.

**Reply by fmw42** [reply]
Zip it and post to some free hosting service and post the URL.

**Reply by fmw42** [reply]
You have gslib in your list of Delegates, so it is installed and IM recognizes it. And your policy permits it and it is properly listed in your delegates.xml file.

Can you process a PDF?  If so, then Ghostscript is installed properly.

If that is the case, then perhaps it is the AI file?

**Reply by DadeMurphy72** [reply]
https://we.tl/t-Vom6g4GXn0 ai file here. I can convert this using Ghostscript, just not ImageMagick.

**Reply by fmw42** [reply]
It converts fine for me on my Mac using IM 7.1.1.34 and GS 10.03

` magick debug.ai debug.png`

![debug](https://github.com/ImageMagick/ImageMagick/assets/13458218/7744a13e-7528-4ade-9d81-71f8427029e7)

**Reply by fmw42** [reply]
Can you use IM to process a PDF?

**Reply by fmw42** [reply]
Is Ghostscript in your PATH environment variable?  Is the path correct in your delegatemap?

**Reply by DadeMurphy72** [reply]
Doesn't seem to be converting PDF's either. It will do jpg.

(new_ENV) D:\Projects\PJC13\resources\ImageMagick7>magick "C:\temp\debug.pdf" "C:\temp\debug.png"
magick: no images for write '-write' 'C:\temp\debug.png' at CLI arg 2 @ error/operation.c/CLINoImageOperator/4983.

**Reply by fmw42** [reply]
If it won't convert PDFs (but Ghostscript by itself will), then it would either be your IM command or your Ghostscript installation in Imagemagick or your IM installation.  Sorry, I do not know what to suggest other than a re-install.

**Reply by DadeMurphy72** [reply]
Thanks. The PATH is correct, I've set them all accordingly.

**Reply by DadeMurphy72** [reply]
When it converted fine for you, is ImageMagick installed on your system or are you running it in an environment?

**Reply by fmw42** [reply]
Installed in my system

**Reply by DadeMurphy72** [reply]
Ok, I can convert with no issues when installed, but again, unfortunately, I cannot install it due to conflicting with the RIP. I'll go back to the portable and try again.

**Comment by DadeMurphy72** [comment, marked as answer]
Ok, there are a couple of things I've come across that "resolves" the issue.

1. Setting Environment Variables
```
 # Incorrect:
    gs_prog = os.path.join(base_path, "resources", "ghostscript", "bin", "gswin64c.exe")
# Correct
    gs_prog = os.path.join(base_path, "resources", "ghostscript", "bin")
```
2. Using Wand - From what I can understand after digging deeper, Wand relies on the static version of ImageMagick. I can't find anywhere that it currently using or in the future will support the portable version. Because of this, I've removed Wand and am working with Subprocess. It's a bit more challenging for what my app does, but at least I have a path forward.

**Comment by Anke** [comment]
Hello,

I set up Xampp Lite (portable) for Kirby CMS. Kirby needs ImageMagick and Imagick to process images. I sucessfully integrated imagick but now I'm at a loss. I copied all files from ImageMagick-7.1.2-9-portable-Q16-x64.7z into \xampp_lite_8_4\apps, not knowing if this is the place it belongs. Since it's portable I guess I need to let Apache and/or PHP know where to find IM. Can you help?

Anke
